pkg/kubelet/status/manager.go

```diff
@@ -323,8 +323,14 @@ func (m *manager) syncPod(uid types.UID, status versionedPodStatus) {
 	// TODO: make me easier to express from client code
 	pod, err := m.kubeClient.Pods(status.podNamespace).Get(status.podName)
 	if errors.IsNotFound(err) {
-		glog.V(3).Infof("Pod %q (%s) was deleted on the server", status.podName, uid)
-		m.deletePodStatus(uid)
+		glog.V(3).Infof("Pod %q (%s) does not exist on the server", status.podName, uid)
+		// Note that we do not delete the pod status because the pod to update may
+		// be a static pod whose mirror pod is in the process of recreation.
+		// Deleting the status without clearing the corresponding
+		// apiStatusVersions entry may lead to wrong versioning and stale status.
```

Comment:
This and https://github.com/kubernetes/kubernetes/issues/8818 would be much simpler if we actually had a proper version on the status object in the API server.

I still think your change fixes the issue you described above.